src/batcher.js

```diff
@@ -253,15 +253,21 @@ class Batcher {
           }
         }
 
+        const savedBatchStreams = this.batch.streams
+        
+        // No need to clear the batch if batching is disabled
+        logEntry === undefined && this.clearBatch()
+
         // Send the data to Grafana Loki
         req.post(this.url, this.contentType, this.options.headers, reqBody, this.options.timeout, this.options.httpAgent, this.options.httpsAgent)
           .then(() => {
-            // No need to clear the batch if batching is disabled
-            logEntry === undefined && this.clearBatch()
             this.batchSent()
             resolve()
           })
           .catch(err => {
+            // Revert batch for the logs to be sent on the next iteration
+            this.batch.streams = savedBatchStreams
```

Comment:
Sorry, took a while to look at this, but doesn't this just reintroduce the same race condition but in a new place? You'd need to combine the savedBatchStreams & this.batch.streams in a non-duplicative manner, like merging sets.

Reply:
This sure fixes the problem so that it only happens when the request fails which decreases its frequency quite a bit. I'll draft something based on this :)